### ISSUE TITLE
It’s not possible to send a SIGTERM signal, which requests the application to gracefully terminate

### ISSUE THREAD

**Issue by PetarRAutoAddress** [dotnet/runtime#59746]
[The .NET Process class on Linux](https://developers.redhat.com/blog/2019/10/29/the-net-process-class-on-linux#the_basics)

So, it seems that on Linux (container) using .NET I can start a new process, but I cannot tell it to gracefully terminate ?

This seems a major problem, that makes any serious process handling on Linux  non-viable.

Any suggestions?

**Comment by PetarRAutoAddress** [comment]
> I couldn't figure out the best area label to add to this issue. If you have write-permissions please help me learn by adding exactly one [area label](https://github.com/dotnet/runtime/blob/master/docs/area-owners.md).

area-System.Diagnostics.Process

**Comment by PathogenDavid** [comment]
I agree this is probably something that the runtime would provide and seems like a natural extension of https://github.com/dotnet/runtime/issues/50527.

That being said if you need something that works today, I've used [Mono.Unix](https://www.nuget.org/packages/Mono.Unix) for this in the past. (Despite the name, it is not Mono-specific.)

Here's an example using it:

```csharp
//using Mono.Unix;
//using Mono.Unix.Native;
//using System;
//using System.Diagnostics;

void SendSigInt(Process process)
{
    if (Syscall.kill(process.Id, Signum.SIGINT) == 0)
    {
        Console.WriteLine("Sent SIGINT successfully.");
    }
    else
    {
        Errno errno = Stdlib.GetLastError();
        Console.WriteLine($"Failed to send SIGINT, errno = {errno} ({UnixMarshal.GetErrorDescription(errno)})");
    }
}
```

**Comment by PetarRAutoAddress** [comment]
Thanks.
Can it send SIGTERM as well ?

**Comment by PathogenDavid** [comment]
> Thanks. Can it send SIGTERM as well ?

Yes, that and [basically every signal you'd ever want](https://github.com/mono/mono.posix/blob/057e4db75dd7d1f51e5409f214c94b91f7849b89/src/Mono.Unix/Mono.Unix.Native/Syscall.cs#L304-L340).

Mono.Unix is a thin wrapper around much of the Unix/POSIX API. So when you call `Sys.kill` you're basically calling [kill(2)](https://manpages.ubuntu.com/manpages/focal/en/man2/kill.2.html) directly.

**Comment by tmds** [comment]
Alternatively, you can:

```cs
using System.Diagnostics;
using System.Runtime.InteropServices;

public static class ProcessExtensions
{
    private const int SIGTERM = 15;

    public static void Terminate(this Process process)
    {
        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
        {
            return;
        }

        if (!process.HasExited)
        {
            kill(process.Id, SIGTERM);
        }
    }

    [DllImport("libc")]
    private static extern int kill(int pid, int sig);
}
```

**Comment by PetarRAutoAddress** [comment]
Great! Thanks.

**Comment by moh-hassan** [comment]
Can this feature to be part of NET9

**Comment by xecrets** [comment]
The suggested interim solution https://github.com/mono/mono.posix doesn't support Windows, while #50527 does. It would be very nice with a .NET API supported across all platforms, within the limitation of the respective platform obviously. But at least the equivalent of SIGINT is possible to generate on Windows, and it should be supported if this issue does get implemented. Currently platform specific code with various kludges are necessary to achieve this, similar to https://github.com/devlooped/dotnet-stop . It would be nice to have a simple supported .NET API for this!

**Comment by adamsitnik** [comment]
We have introduced an API that allows to signal the Process in .NET 11 Preview 4, more details and code samples: https://devblogs.microsoft.com/dotnet/process-api-improvements-in-dotnet-11/#safeprocesshandle

There is now also an ability to get the terminating signal of given process.
